checker/src/org/checkerframework/checker/upperbound/UpperBoundAnnotatedTypeFactory.java

```diff
@@ -0,0 +1,770 @@
+                return true;
+            } else if (AnnotationUtils.areSameByClass(rhs, LTEqLengthOf.class)) {
+                return false;
+            }
+
+            // Neither is LTEL. Both must be LTOM, LTL, or Bottom. And the two must be
+            // different. The only way this can return true is if rhs is Bottom, or if rhs is
+            // LTOM and lhs is LTL.
+            if (AnnotationUtils.areSameByClass(rhs, UpperBoundBottom.class)) {
+                return true;
+            }
+
+            if (AnnotationUtils.areSameByClass(rhs, LTOMLengthOf.class)
+                    && AnnotationUtils.areSameByClass(lhs, LTLengthOf.class)) {
+                return true;
+            }
+
+            // Every other case results in false.
+            return false;
+        }
+    }
+
+    @Override
+    public TreeAnnotator createTreeAnnotator() {
+        return new ListTreeAnnotator(
+                new UpperBoundTreeAnnotator(this), super.createTreeAnnotator());
+    }
+
+    protected class UpperBoundTreeAnnotator extends TreeAnnotator {
+
+        public UpperBoundTreeAnnotator(UpperBoundAnnotatedTypeFactory factory) {
+            super(factory);
+        }
+
+        /**
+         * This exists specifically for Math.min. We need to special case Math.min because it has
+         * unusual semantics: it can be used to combine annotations for the UBC, so it needs to be
+         * special cased. Do not special case other methods here unless you have a compelling reason
+         * to do so.
+         */
+        @Override
+        public Void visitMethodInvocation(MethodInvocationTree tree, AnnotatedTypeMirror type) {
+            ExecutableElement fcnMin =
+                    TreeUtils.getMethod("java.lang.Math", "min", 2, processingEnv);
+            if (TreeUtils.isMethodInvocation(tree, fcnMin, processingEnv)) {
+                AnnotatedTypeMirror leftType = getAnnotatedType(tree.getArguments().get(0));
+                AnnotatedTypeMirror rightType = getAnnotatedType(tree.getArguments().get(1));
+
+                type.replaceAnnotation(
+                        qualHierarchy.greatestLowerBound(
+                                leftType.getAnnotationInHierarchy(UNKNOWN),
+                                rightType.getAnnotationInHierarchy(UNKNOWN)));
+            }
+            return super.visitMethodInvocation(tree, type);
+        }
+
+        @Override
+        public Void visitMemberSelect(MemberSelectTree tree, AnnotatedTypeMirror type) {
+            if (tree.getIdentifier().contentEquals("length")
+                    && InternalUtils.typeOf(tree.getExpression()).getKind() == TypeKind.ARRAY) {
+                String arrName =
+                        FlowExpressions.internalReprOf(this.atypeFactory, tree.getExpression())
+                                .toString();
+                type.replaceAnnotation(
+                        qualHierarchy.greatestLowerBound(
```

Comment:
The annotated type of a field access is computed as follows: 
1. The type of the declaration of the field is computed. 
    A. Annotations explicitly written on the element are applied. (Array.length can't be explicitly annotated, so nothing is applied.)
    B. Annotations on explicitly written on declaration of the type of the field are applied. (The type of `length` is int, so there are no annotations.)
    C. The TypeAnnotator is applied on the type. (The TypeAnnotator doesn't do anything with ints, so no annotations are applied.)
    D. Defaults are applied.  `length` is a field so, it is defaulted to top.
2. The type is viewpointed adapted based on the type of the expression that access the field.  (There is not viewpoint adaption for Arrays, so this doesn't effect the type, so it is still annotated with top.)
3. The TreeAnnotator is applied.  (The annotation on type is top and can only be top.)
4. The TypeAnnotator is applied.
5. Defaults are applied.
6. Annotations from data flow are applied. (If the annotation from dataflow is a subtype of the annotation on type, then the dataflow annotation is applied.  Otherwise, type is unchanged.)

So, unless the default annotation for fields is changed, then the annotation on `type` at this point can only be `@UpperBoundUnknown`.